bip-0341.mediawiki

```diff
@@ -284,6 +284,111 @@ The reason for this is to increase leaf entropy and prevent an observer from lea
 
 == Test vectors ==
 
+=== SIGHASH ALL ===
+  
+  The following is an unsigned transaction:
+    02000000025f6092ec9bb430830dfc344260dd5a03cf355186e774be49b2fe5c362f56cb8d00000000000000000061431892d76aa28b5ed1e3da8800fa0d7190c4b4f22be5f416d2d07e573b32e10100000000000000000100ca9a3b000000001976a914682dfdbc97ab5c31300f36d3c12c6fd854b1b35a88ac00000000
+  
+  The first input comes from a P2WPKH
+    scriptPubKey : 0014196a5bea745288a7f947993c28e3a0f2108d2e0a value: 5.0
+    private key  : 6b3973ee2ce444ada0147716925f6f77569350804835498593dd3be95163d558
+    public key   : 0271be339aeae9ed2c6a5a7f8ac5f49638da387612be881c7ed2fb3848b0ef8a6c
+    
+  The second input comes from a P2TR
+    scriptPubKey : 5120cdf7dec1d4282be1bcca13f3f7cf5befec97c68911fd5a3b461aac419557fd17, value: 6
+    private key  : cf3780a32ef3b2d70366f0124ee40195a251044e82a13146106be75ee049ac02
+    
+  To sign it with a nHashType of SIGHASH_ALL:
+  
+  shaPrevouts:
+    SHA256(5f6092ec9bb430830dfc344260dd5a03cf355186e774be49b2fe5c362f56cb8d0000000061431892d76aa28b5ed1e3da8800fa0d7190c4b4f22be5f416d2d07e573b32e101000000)
+  = 32553b113292dfa8216546e721388a6c19c76626ca65dc187e0348d6ed445f81
+  
+  shaAmounts:
+    SHA256(0065cd1d000000000046c32300000000)
+  = 5733468db74734c00efa0b466bca091d8f1aab074af2538f36bd0a734a5940c5
+  
+  shaScriptpubkeys:
+    SHA256(160014196a5bea745288a7f947993c28e3a0f2108d2e0a225120cdf7dec1d4282be1bcca13f3f7cf5befec97c68911fd5a3b461aac419557fd17)
+  = 189f262af17538f5e5df96b574a51c29038341001b8fd5ae356f985b9b880f1f
+  
+  shaSequences:
+    SHA256(0000000000000000)
+  = af5570f5a1810b7af78caf4bc70a660f0df51e42baf91d4de5b2328de0e83dfc
+  
+  shaOutputs:
+    SHA256(00ca9a3b000000001976a914682dfdbc97ab5c31300f36d3c12c6fd854b1b35a88ac)
+  = 8cdee56004a241f9c79cc55b7d79eaed04909d84660502a2d4e9c357c2047cf5
+  
+  hash preimage: 0000020000000000000032553b113292dfa8216546e721388a6c19c76626ca65dc187e0348d6ed445f815733468db74734c00efa0b466bca091d8f1aab074af2538f36bd0a734a5940c5189f262af17538f5e5df96b574a51c29038341001b8fd5ae356f985b9b880f1faf5570f5a1810b7af78caf4bc70a660f0df51e42baf91d4de5b2328de0e83dfc8cdee56004a241f9c79cc55b7d79eaed04909d84660502a2d4e9c357c2047cf50001000000
+    
+  sigHash:      71cc53d93fa4419e6ea19eeacd4c6ad57d82ee5dee8810cfa5b1a91efbd48329
+  signature:    f6bfa62fd9edb4b823e35d9e72043ec0709fda7ee836d9acb9ebc9709746bc87c67ec4d32e08c92f16aa2e52bbff034d414d89d9e10d5ccc2785acd021c893f8
```

Comment:
This test vector should mention the k nonce derivation follows libsecp256k1, or specify an aux_data for use with nonce generation (because the reference implementation requires it, if we eventually try to move this vector into the reference implementation, it will be needed anyways.)

---

As mentioned in my previous PR (#1218), many people will look at the reference implementation (which requires aux_data) and write their apps in a way that requires aux_data. Whereas these test vectors are using a null pointer for aux_data when signing using libsecp256k1, so many people who don't use libsecp256k1 will be confused.

A simple one-liner explaining the nonce generation using terminology from BIP340 to describe this behavior, OR including a 32 byte aux_data in the vector (which is probably simpler) should be done in my opinion.

Other than that, LGTM, thanks a ton!

Reply:
Hi, thank for your suggestions. Including a 32 bytes aux_data instead of an unknown random one would certainly help because the entire test would be reproducible, but this is not its main objective. The main objective is to test the signature message, not the signature itself. Anyway if I am not mistaken we are going to close this PR, and accept the test vector by @sipa which already do what you suggested. 

But for the rest I have to disagree with you. As people in your previous PR have already pointed out, a BIP is not tied to any specific library. It only provides a link to a reference implementation which may or may not be the main one (bitcoin core or libsecp256k1). That's because a BIP is a design document, the reference is only there to help people who want to implement the BIP.

Reply:
The new vector containing aux_rand will make things much clearer to the implementer. I'm glad it will be included.

---

I never claimed that a BIP should be tied to any library. In the other PR I merely stated that having a reference implementation that takes one specific path on an aspect which the design document states is optional (and has infinite possible paths) without any mechanism to show it's optional (ie. make it an option, pass the nonce function as an argument to sign, or even just write a comment in the reference reminding people "this is explicitly diverging from the design document for simplicity's sake and implementing the recommended nonce derivation function which requires aux_data, while recommended, this is strictly optional" etc.) will (and has) lead to confusion.

My proposal to fix that point of possible confusion was implicitly rejected, and I really don't care too much to fight for my specific fix.

However, debating the point that it is not confusing to anyone I would be more adamant.

That said, this debate is not something to be had on this PR.

Thanks for your time.

Reply:
@junderw https://github.com/bitcoin-core/secp256k1/pull/1002